common/src/main/java/com/microsoft/identity/common/internal/logging/CommonCoreLogger.java

```diff
@@ -0,0 +1,223 @@
+package com.microsoft.identity.common.internal.logging;
+
+import android.os.Build;
+import android.util.Log;
+
+import com.microsoft.identity.common.adal.internal.util.StringExtensions;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.TimeZone;
+import java.util.concurrent.atomic.AtomicReference;
+
+public final class CommonCoreLogger {
+    private static final CommonCoreLogger INSTANCE = new CommonCoreLogger();
+    static final String DATE_FORMAT = "yyyy-MM-dd HH:mm:ss";
+
+    // Turn on the VERBOSE level logging by default.
+    private LogLevel mLogLevel  = LogLevel.VERBOSE;
+
+    private AtomicReference<ILoggerCallback> mExternalLogger = new AtomicReference<>(null);

```

Comment:
AtomicReference is preferred than synchronized. I am worried if we use synchronized calls on mExternalLogger, then the entire log() function has to be synchronized which is high in the perspective of time cost.

Reply:
If the goal is to disallow overwriting an already set `ILoggerCallback` instance, I suspect you'll want to use `compareAndSet(V, V)` rather than separate calls to `get()` and `set()` -- I'm skeptical of `AtomicReference` here, though: I'm not seeing the utility....

If we only expect a developer to call `setExternalLogger` once, it seems like this method could be `synchronized` and subsequent calls would fail with an `IllegalStateException` -- is there something else I'm missing? We're [initializing a singleton](https://stackoverflow.com/questions/11165852/java-singleton-and-synchronization), right?